dkg/state_machine.go

```diff
@@ -0,0 +1,700 @@
+var ErrDuplicateRejection = errors.New("this participant already rejected the proposal")
+var ErrNonLeaderCannotReceiveAcceptance = errors.New("you received an acceptance but are not the leader of this DKG - cannot do anything")
+var ErrNonLeaderCannotReceiveRejection = errors.New("you received a rejection but are not the leader of this DKG - cannot do anything")
+var ErrFinalGroupCannotBeEmpty = errors.New("you cannot complete a DKG with a nil final group")
+var ErrKeyShareCannotBeEmpty = errors.New("you cannot complete a DKG with a nil key share")
+
+//nolint:gocyclo
+func isValidStateChange(current, next DKGStatus) bool {
+	switch current {
+	case Complete:
+		return next == Proposing || next == Proposed
+	case Aborted:
+		return next == Proposing || next == Proposed
+	case TimedOut:
+		return next == Proposing || next == Proposed || next == Evicted
+	case Fresh:
+		return next == Proposing || next == Proposed
+	case Joined:
+		return next == Left || next == Executing || next == Aborted || next == TimedOut
+	case Left:
+		return next == Joined || next == Aborted
+	case Proposing:
+		return next == Executing || next == Aborted || next == TimedOut
+	case Proposed:
+		return next == Accepted || next == Rejected || next == Aborted || next == TimedOut || next == Left || next == Joined
+	case Accepted:
+		return next == Executing || next == Aborted || next == TimedOut
+	case Rejected:
+		return next == Aborted || next == TimedOut
+	case Executing:
+		return next == Complete || next == TimedOut || next == Evicted
+	case Evicted:
+		return next == Joined
+	}
+	return false
+}
+
+func hasTimedOut(details *DBState) bool {
+	now := time.Now()
+	return details.Timeout.Before(now) || details.Timeout.Equal(now)
+}
+
+func ValidateProposal(currentState *DBState, terms *drand.ProposalTerms) error {
+	if currentState.BeaconID != terms.BeaconID {
+		return ErrInvalidBeaconID
+	}
+
+	_, found := crypto.GetSchemeByID(terms.SchemeID)
+	if !found {
+		return ErrInvalidScheme
+	}
+
+	if terms.Timeout.AsTime().Before(time.Now()) {
+		return ErrTimeoutReached
+	}
+
+	if int(terms.Threshold) > len(terms.Joining)+len(terms.Remaining) {
+		return ErrThresholdHigherThanNodeCount
+	}
+
+	if err := validateEpoch(currentState, terms); err != nil {
+		return err
+	}
+
+	if terms.TransitionTime == nil {
+		return ErrTransitionTimeMissing
+	}
+
+	if terms.Epoch == 1 {
+		return validateFirstEpoch(terms)
+	}
+
+	// perhaps this should be stricter?
+	// should there be at least one round? should it be after `time.Now()`?
+	if terms.TransitionTime.AsTime().Before(currentState.GenesisTime) {
+		return ErrTransitionTimeBeforeGenesis
+	}
+
+	if util.Contains(terms.Joining, terms.Leader) {
+		return ErrLeaderCantJoinAfterFirstEpoch
+	}
+
+	if len(terms.Remaining) == 0 {
+		return ErrNoNodesRemaining
+	}
+
+	if util.Contains(terms.Leaving, terms.Leader) || !util.Contains(terms.Remaining, terms.Leader) {
```

Comment:
Here we clearly see that the dbstate contains redundant information and that you need to check for both. Remaining = previous nodes - leaving, but yet it's a separate list. 
I guess it's okay but it's also fragile, need to maintain two state that represent the same information in sync.